src/Params/WaveTable.h

```diff
@@ -67,40 +66,153 @@ class Shape
     }
 };
 
-//! Common Tensor base class for all dimensions
+/**
+    Common Tensor base class for all dimensions
+
+    m_usable <= m_size_planned <= m_capacity
+    m_usable: data that has been generated and can be used
+    m_size_planned: data that shall be generated (after generation,
+                    m_usable == m_size_planned)
+    m_capacity: allocated size
+*/
 template <std::size_t N, class T>
 class TensorBase
 {
 protected:
     std::size_t m_capacity;
-    bool     m_owner = true; //! says if memory is owned by us
+    int m_size;
+    bool m_owner = true; //!< says if memory is owned by us
 
-    TensorBase() : m_capacity(0), m_owner(false) {}
-    TensorBase(std::size_t capacity) : m_capacity(capacity), m_owner(true) {};
+    TensorBase() : m_capacity(0), m_size(0), m_owner(false) {}
+    TensorBase(std::size_t capacity) : m_capacity(capacity), m_size(0), m_owner(true) {};
 
+#if 0
     TensorBase(TensorBase&& other) {
         m_capacity = other.capacity;
+        m_size_planned = other.m_size_planned;
+        m_usable = other.m_usable;
         m_owner = other.m_owner;
         other.m_owner = false;
     }
     TensorBase& operator=(TensorBase&& other) {
         m_capacity = other.m_capacity;
+        m_size_planned = other.m_size_planned;
+        m_usable = other.m_usable;
         m_owner = other.m_owner;
         other.m_owner = false;
     }
+    // TODO: check rule of 5
+#endif
 public:
     std::size_t capacity() const { return m_capacity; }
+    void resize(int newsize) { m_size = newsize; }
+    int size() const { return m_size; }
 protected:
-    void reserve_capacity(std::size_t new_capacity) { m_capacity = new_capacity; m_owner = true; }
+    void set_capacity(std::size_t new_capacity) { m_capacity = new_capacity; m_owner = true; }
     bool operator==(const TensorBase<N, T>& other) const {
 #if 0
         return shape() == other.shape();/* &&
                 std::equal(m_data, m_data + (shape().volume()), other.m_data);*/ }
 #else
-        return m_capacity == other.m_capacity;
+        return m_capacity == other.m_capacity && m_size == other.m_size;
+#endif
+    }
+
+
+    void swapWith(Tensor<N,T>& other)
+    {
+        std::swap(m_capacity, other.m_capacity);
+        std::swap(m_size, other.m_size);
+        std::swap(m_owner, other.m_owner);
+    }
+};
+
+//! Ringbuffer without buffer - only size, reader and writer
+class AbstractRingbuffer
```

Comment:
Based on the timing diagram it sounds like everything is worked on within the same thread. If that's the case you don't need to worry about atomics.

Reply:
Exactly. The atomics are just remaining out-if-deffed code from earlier, I did not clean that up properly.

Reply:
No problem. First time I passed through the code I was thinking there was an alternative codepath to directly have middleware and the backend talking, which appears to be me way overcomplicating what you actually did.